src/token/cose/encrypted/encrypt0/mod.rs

```diff
@@ -0,0 +1,107 @@
+use alloc::rc::Rc;
+use alloc::vec::Vec;
+use core::cell::RefCell;
+
+use coset::{CoseEncrypt0, CoseEncrypt0Builder, EncryptionContext, Header};
+
+use crate::error::CoseCipherError;
+use crate::token::cose::encrypted;
+use crate::token::cose::encrypted::CoseEncryptCipher;
+use crate::token::cose::key::{CoseAadProvider, CoseKeyProvider};
+
+#[cfg(all(test, feature = "std"))]
+mod tests;
+
+pub trait CoseEncrypt0Ext {
+    fn try_decrypt<B: CoseEncryptCipher, CKP: CoseKeyProvider, CAP: CoseAadProvider>(
+        &self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        external_aad: &mut CAP,
+    ) -> Result<Vec<u8>, CoseCipherError<B::Error>>;
+}
+
+impl CoseEncrypt0Ext for CoseEncrypt0 {
+    fn try_decrypt<B: CoseEncryptCipher, CKP: CoseKeyProvider, CAP: CoseAadProvider>(
+        &self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        external_aad: &mut CAP,
+    ) -> Result<Vec<u8>, CoseCipherError<B::Error>> {
+        let backend = Rc::new(RefCell::new(backend));
+        let key_provider = Rc::new(RefCell::new(key_provider));
+        self.decrypt(
+            external_aad.lookup_aad(
+                Some(EncryptionContext::CoseEncrypt0),
+                Some(&self.protected.header),
+                Some(&self.unprotected),
+            ),
+            |ciphertext, aad| {
+                encrypted::try_decrypt(
+                    &backend,
+                    &key_provider,
+                    &self.protected.header,
+                    &self.unprotected,
+                    try_all_keys,
+                    ciphertext,
+                    aad,
+                )
+            },
+        )
+    }
+}
+
+pub trait CoseEncrypt0BuilderExt: Sized {
+    fn try_encrypt<B: CoseEncryptCipher, CKP: CoseKeyProvider, CAP: CoseAadProvider>(
+        self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
+        protected: Option<Header>,
+        unprotected: Option<Header>,
+        plaintext: &[u8],
+        external_aad: &mut CAP,
+    ) -> Result<Self, CoseCipherError<B::Error>>;
+}
+
+impl CoseEncrypt0BuilderExt for CoseEncrypt0Builder {
+    fn try_encrypt<B: CoseEncryptCipher, CKP: CoseKeyProvider, CAP: CoseAadProvider>(
+        self,
+        backend: &mut B,
+        key_provider: &mut CKP,
+        try_all_keys: bool,
```

Comment:
A more general question: Would it be preferable to include the `try_all_keys` field in the key provider in some way?

For instance, one could implement `CoseKeyProvider` such that the key_id is entirely disregarded for the basic types (equivalent to `try_all_keys == true)`, and then provide a wrapper `FilterMatchingKeyIds` that takes another `CoseKeyProvider` and modifies its output such that only matching key IDs are returned (equivalent to `try_all_keys == false`).

That way, we could save an argument here that will probably not be relevant to most users (as most will probably just provide a single key or a slice of keys anyways).

Another thing regarding the key providers: Should we maybe only provide a immutable reference to the key provider in the lookup method instead of a mutable one?
Most `CoseKeyProvider`s will not require mutable access anyways, and it would get rid of all of these weird instances of `&mut &key` all throughout the code.
If mutable access is required for a key provider, it could still use interior mutability (using a `RefCell`/`Mutex`).

Reply:
> A more general question: Would it be preferable to include the try_all_keys field in the key provider in some way?

Yes, I think that makes sense, given that most people won't need to use this parameter anyway.

> Another thing regarding the key providers: Should we maybe only provide a immutable reference to the key provider in the lookup method instead of a mutable one?

To be honest, I'm not familiar enough with the interior mutability stuff in Rust to provide a confident answer here. Instead, I can only give the rather unhelpful answer that it depends on how many key providers actually need a mutable reference, and how much the switch to interior mutability would complicate the provider's implementation.

Reply:
> Yes, I think that makes sense, given that most people won't need to use this parameter anyway.

Okay, I will adjust the implementation accordingly.

> [...] it depends on how many key providers actually need a mutable reference, and how much the switch to interior mutability would complicate the provider's implementation.

For the provider implementation, if you need mutable access you would basically have to create a wrapper struct with a `RefCell<T>` field, where `T` is your actual key provider.
When calling `lookup_key()`, you'd then have to call `RefCell::borrow_mut()` to get a mutable reference and then call the actual lookup method.

For most basic key containers (`Vec`s, slices, `HashMap`, `BTreeMap`, ...) a simple immutable reference should always suffice. 
A mutable reference is only needed if you actually need to change something in the key provider in order to retrieve a key. One example for that might be database backed key providers, where you might have some internal state that has to be modified to perform the query.

Reply:
Ah, so that's basically runtime borrow-checking, right? I think it's fine to make it non-mut, then. It sounds to me like most key providers are simple enough that an immutable reference suffices, and for others, the `RefCell` solution is still a reasonable fallback.